wfe2/verify.go

```diff
@@ -192,28 +212,55 @@ func (wfe *WebFrontEndImpl) validNonce(ctx context.Context, header jose.Header)
 		wfe.stats.joseErrorCount.With(prometheus.Labels{"type": "JWSMissingNonce"}).Inc()
 		return probs.BadNonce("JWS has no anti-replay nonce")
 	}
+	var noBackend bool
 	var valid bool
 	var err error
 	if wfe.noncePrefixMap == nil {
 		// Dispatch nonce redemption RPCs dynamically.
+		prob := nonceWellFormed(header.Nonce, nonce.PrefixLen)
+		if prob != nil {
+			wfe.stats.joseErrorCount.With(prometheus.Labels{"type": "JWSMalformedNonce"}).Inc()
+			return prob
+		}
+
 		ctx = context.WithValue(ctx, nonce.PrefixCtxKey{}, header.Nonce[:nonce.PrefixLen])
 		ctx = context.WithValue(ctx, nonce.HMACKeyCtxKey{}, wfe.rncKey)
 		resp, err := wfe.rnc.Redeem(ctx, &noncepb.NonceMessage{Nonce: header.Nonce})
 		if err != nil {
-			return web.ProblemDetailsForError(err, "failed to redeem nonce")
+			rpcStatus, ok := status.FromError(err)
+			if !ok {
+				return web.ProblemDetailsForError(err, "failed to redeem nonce")
+			}
+			if rpcStatus == nb.ErrNoBackendsMatchPrefix {
+				// ErrNoBackendsMatchPrefix indicates the provided prefix is
+				// unknown to the nonce redemption service. This is a transient
+				// failure, the client should retry with a new nonce.
+				resp = &noncepb.ValidMessage{Valid: false}
+				noBackend = true
+			}
```

Comment:
This doesn't handle the case of `err != nil` but `rpcStatus != nb.ErrNoBackendsMatchPrefix`.